dbms/src/Storages/S3/FileCache.cpp

```diff
@@ -0,0 +1,552 @@
+    }
+    return true;
+}
+
+void FileCache::downloadImpl(const String & s3_key, FileSegmentPtr & file_seg)
+{
+    Stopwatch sw;
+    auto client = S3::ClientFactory::instance().sharedClient();
+    const auto & bucket = S3::ClientFactory::instance().bucket();
+    Aws::S3::Model::GetObjectRequest req;
+    req.SetBucket(bucket);
+    req.SetKey(s3_key);
+    ProfileEvents::increment(ProfileEvents::S3GetObject);
+    auto outcome = client->GetObject(req);
+    if (!outcome.IsSuccess())
+    {
+        throw S3::fromS3Error(outcome.GetError(), "s3_key={}", s3_key);
+    }
+    auto & result = outcome.GetResult();
+    auto content_length = result.GetContentLength();
+    ProfileEvents::increment(ProfileEvents::S3ReadBytes, content_length);
+    GET_METRIC(tiflash_storage_s3_request_seconds, type_get_object).Observe(sw.elapsedSeconds());
+    if (!finalizeReservedSize(file_seg->getFileType(), file_seg->getSize(), content_length))
+    {
+        LOG_DEBUG(log, "s3_key={} finalizeReservedSize {}=>{} failed.", s3_key, file_seg->getSize(), content_length);
+        return;
+    }
+    file_seg->setSize(content_length);
+
+    const auto & local_fname = file_seg->getLocalFileName();
+    prepareParentDir(local_fname);
+    auto temp_fname = toTemporaryFilename(local_fname);
+    {
+        Aws::OFStream ostr(temp_fname, std::ios_base::out | std::ios_base::binary);
+        RUNTIME_CHECK_MSG(ostr.is_open(), "Open {} failed: {}", temp_fname, strerror(errno));
+        ostr << result.GetBody().rdbuf();
+        RUNTIME_CHECK_MSG(ostr.good(), "Write {} failed: {}", temp_fname, strerror(errno));
+        ostr.flush();
+    }
+    std::filesystem::rename(temp_fname, local_fname);
+    file_seg->setStatus(FileSegment::Status::Complete);
+    LOG_DEBUG(log, "Download s3_key={} to local={} size={} cost={}ms", s3_key, local_fname, content_length, sw.elapsedMilliseconds());
+}
+
+void FileCache::download(const String & s3_key, FileSegmentPtr & file_seg)
+{
+    try
+    {
+        downloadImpl(s3_key, file_seg);
+    }
+    catch (...)
+    {
+        tryLogCurrentException(log, fmt::format("Download s3_key={} failed", s3_key));
+    }
+
+    if (!file_seg->isReadyToRead())
+    {
+        bg_download_fail_count.fetch_add(1, std::memory_order_relaxed);
+        file_seg.reset();
+        remove(s3_key);
+    }
+    else
+    {
+        bg_download_succ_count.fetch_add(1, std::memory_order_relaxed);
+    }
+    bg_downloading_count.fetch_sub(1, std::memory_order_relaxed);
+    LOG_DEBUG(log, "downloading count {} => s3_key {} finished", bg_downloading_count.load(std::memory_order_relaxed), s3_key);
+}
+
+void FileCache::bgDownload(const String & s3_key, FileSegmentPtr & file_seg)
+{
+    bg_downloading_count.fetch_add(1, std::memory_order_relaxed);
+    LOG_DEBUG(log, "downloading count {} => s3_key {} start", bg_downloading_count.load(std::memory_order_relaxed), s3_key);
+    IOThreadPool::get().scheduleOrThrowOnError(
```

Comment:
Do we need some protection about too many download tasks that `IOThreadPool` will never catch up?  Maybe we need a limit of max waiting tasks.